api/provider/aws/task_create.go

```diff
@@ -5,29 +5,63 @@ import (
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/ecs"
+	"github.com/quintilesims/layer0/common/config"
 	"github.com/quintilesims/layer0/common/models"
 )
 
-// Create runs an ECS Task using the specified Create Task Request. The Create Task
-// Request contains the Task name, the Environment ID, and the Deploy ID.
-// The Deploy ID is used to look up the ECS Task Definition family and version of the
-// Task to run.
+// Create runs an ECS Task using the specified CreateTaskRequest.
+// The CreateTaskRequest contains the ContainerOverrides, the DeployID, the
+// EnvironmentID, the TaskName, and the TaskType.
+//
+// The Deploy ID is used to look up the ECS TaskDefinition Family and Version
+// of the Task to run. The TaskType parameter is one of "stateful" or "stateless"
+// and indicates which ECS LaunchType the user wishes to use ("EC2" or "FARGATE"
+// respectively).
+//
+// Create does not generate any custom errors of its own, but will bubble up errors
```

Comment:
It looks like the service create retry logic is swallowing all errors right now:
```
	fn := func() (shouldRetry bool, err error) {
		if err := s.createService(
			cluster,
			serviceName,
			taskDefinitionARN,
			loadBalancerRole,
			networkMode,
			req.Stateful,
			scale,
			subnets,
			securityGroupIDs,
			loadBalancer,
		); err != nil {
			if strings.Contains(err.Error(), "Unable to assume role") {
				log.Printf("[DEBUG] Failed service create, will retry (%v)", err)
				return true, nil
			}

			return false, err
		}

		return false, nil
	}

	if err := retry.Retry(fn, retry.WithTimeout(time.Second*30), retry.WithDelay(time.Second)); err != nil {
		return "", errors.New(errors.EventualConsistencyError, err)
	}
```

In a situation where `fn()` returns `false` (for instance, if we obtain an AWS error because the user is trying to create a stateless service using a stateful deploy), the call is made three times and then an EventualConsistencyError is returned.

Reply:
| SVC Type | LB Type | DPL Type | Result |
|---|---|---|---|
| stateless | alb | fargate | ✅ |
| stateless | alb | ec2 | AWS `InvalidParameterException`: task def does not support fargate |
| stateless | clb | fargate | Layer0 error: unsupported stateless+clb |
| stateless | clb | ec2 | Layer0 error: unsupported stateless+clb |
| stateless | --- | fargate | ✅ |
| stateless | --- | ec2 |  |
| stateful | alb | fargate | Layer0 error: unsupported stateful+alb |
| stateful | alb | ec2 | Layer0 error: unsupported stateful+alb |
| stateful | clb | fargate | AWS `InvalidParamaterException`: task def networkMode awsvpc incompatible with clb |
| stateful | clb | ec2 | ✅ |
| stateful | --- | fargate |  |
| stateful | --- | ec2 | ✅ |

Reply:
✅